lib/midiHandler.py

```diff
@@ -0,0 +1,295 @@
+            else:
+                self.all_unselected_parts.append(part_index)
+                self.selected_instruement_parts[instrument_index] = []
+
+            prev_instrument=part.partId
+
+        print("all_selected_parts = ")
+        print(self.all_selected_parts)
+        print("all_unselected_parts = ")
+        print(self.all_unselected_parts)
+        print("selected_instruement_parts = ")
+        print(self.selected_instruement_parts)
+
+        #play together - all / selected / unselected instruments
+        bpi = int(self.queryString.get("bpi"))
+        self.play_together_unselected = bpi&1
+        bpi=bpi>>1
+        self.play_together_selected = bpi&1
+        bpi=bpi>>1
+        self.play_together_all = bpi&1
+        bpi=bpi>>1
+        
+        
+        while (bsi>1):
+            print ("bsi = " + str(bsi))
+            if (bsi&1==True):
+                self.selected_instruments.append(True)
+            else:
+                self.selected_instruments.append(False)
+            bsi=bsi>>1
+        
+    
+    def make_midi_files(self):
+        #todo - it is a bit slow for big musicxml files eg chaconne!
+        #todo maybe get the segment of all parts then get parts from that.  
+        xml_file_path = os.path.join(*(MEDIA_ROOT, self.folder, self.filename)) #todo - might not be secure 
+        self.score = converter.parse(xml_file_path+".musicxml") #todo - might be .xml instead of .musicxml
+        self.get_selected_instruments()
+            
+        s = stream.Score(id='temp')
+        
+        if self.queryString.get("start") is None and self.queryString.get("end") is None:
+            #todo - test for pickup bar
+            start = self.score.parts[0].getElementsByClass('Measure')[0].number
+            end = self.score.parts[0].getElementsByClass('Measure')[-1].number
+
+        else:
+            start = int(self.queryString.get("start"))
+            end = int(self.queryString.get("end"))
+
+        offset = self.score.parts[0].measure(start).offset
+        
+        #in very rough tests (same bars) using the segment then getting parts from that (instead of original stream) saves 1 or 2 seconds with 2 parts 261 bars, 2,060kb.  
+        #maybe a deep copy would be better?
+        self.scoreSegment = stream.Score(id='tempSegment')
+        for p in self.score.parts:
+            #fix with pickup bar
+            if start==0 and end==0:
+                end=1
+            for m in p.measures(start, end).getElementsByClass('Measure'):
+                #todo - test with repeats
+                m.removeByClass('Repeat') 
+            self.scoreSegment.insert(p.measures(start,end, ))
+            if start==0 and end==1:
+                end=0
+        
+        for click in ['n', 'be']:
+            self.tempo_shift = 0 #in a pickup bar - a rest is added to the start of bar 0 to make it a full bar - this offset needs adding to all future tempos 
```

Comment:
Consider taking out the middle of very long for loop blocks and turning it into a function.

Reply:
Yes - there are some rather long sections there.  As long as I name the functions well - it will make the code easier to read!

Reply:
Fixed in 79fe761
I initially moved parts out into separate functions - then decided the functions were quite similar - so made them just one and added extra parameters.  I need to get the balance between lots of similar functions and one really complicated function used in several places for slightly different purposes...